### PR TITLE
Pinecone batch embed

### PR DESCRIPTION
**Pull request by jamescalam**
Modified the pinecone `add_texts` code to allow an optional `batch_size` parameter. 

If no `batch_size` parameter is given, a default value of `None` will be used and the function will work as it has done in the past.

If a `batch_size` value is passed, the embedding process will take batches of text of size `batch_size`. So rather than embedding one text at a time (current behavior), we can embed 100 or so at a time — significantly speeding up the function.

When using `batch_size` this does mean the user must pass an embedding function that handles batches, this is noted in the doc string. That means that before the initialization of the class using an OpenAI embedding model (`embed`) would have looked like:

```python
vectorstore = Pinecone(index, embed.embed_query,"text")
```

It would now use the document embedding method:

```python
vectorstore = Pinecone(index, embed.embed_documents,"text")
```

### CODE REVIEW

File: langchain/vectorstores/pinecone.py [hunk 1]
```diff
@@ -65,19 +66,27 @@ def add_texts(
...
             List of ids from adding the texts into the vectorstore.
 
         """
         # Embed and create the documents
-        docs = []
+        docs: list = []
         ids = ids or [str(uuid.uuid4()) for _ in texts]
+        # If a batch_size value is provided, we will batch the embeddings
+        if batch_size is not None:
+            embeddings = self._embedding_function(texts)
```
Comment (hwchase17) [inline langchain/vectorstores/pinecone.py:81]: do you have an example of how you would use this? most walkthroughs in the docs use Embedings.embed_query as the function, which takes in a single doc

**Comment by jamescalam** [comment]
@hwchase17 converting to draft — I realize this needs some work, it is possible to do what I was proposing here (batched embeds and upserts) with the `from_texts` method — but it is still very slow compared to doing it directly. So I will take a closer look at what is happening and adjust this PR based on that.

Thanks!

**Comment by hwchase17** [comment]
> @hwchase17 converting to draft — I realize this needs some work, it is possible to do what I was proposing here (batched embeds and upserts) with the `from_texts` method — but it is still very slow compared to doing it directly. So I will take a closer look at what is happening and adjust this PR based on that.
> 
> Thanks!

sg!

**Comment by jamescalam** [comment]
Added changes to this, main thing from a user perspective is that you don't add the embedding method at vector store init, but the embedding object. [Here's an example notebook](https://gist.github.com/jamescalam/6435c4925a3cbf9328e8b566e0969f2d) showing how and where to use this.

Here's the breaking change mentioned above:

```python
# old init code
from langchain.vectorstores import Pinecone

text_field = "text"

vectorstore = Pinecone(
    index, embeddings, text_field
)

# ----
# new init code
from langchain.vectorstores import Pinecone

text_field = "text"

vectorstore = Pinecone(
    index, embeddings.embed_query, text_field
)
```

**Comment by hwchase17** [comment]
thanks - going to think about how to add a more user friendly warning/error to reflect the breaking change...

**Comment by jamescalam** [comment]
@hwchase17 could we add a check for users using the old method? If using that we could add a method like `_query_embed` to handle it and display a deprecation warning? I haven't thought about this too much, but could try and put something together if you approve

File: langchain/vectorstores/pinecone.py [hunk 2]
```diff
@@ -29,13 +33,13 @@ class Pinecone(VectorStore):
...
             embeddings = OpenAIEmbeddings()
-            vectorstore = Pinecone(index, embeddings.embed_query, "text")
+            vectorstore = Pinecone(index, embeddings, "text")
     """
 
     def __init__(
         self,
         index: Any,
-        embedding_function: Callable,
+        embedding: Union[Embeddings, Callable],
```
Comment (dev2049) [inline langchain/vectorstores/pinecone.py:42]: @hwchase17 thoughts on this for backwards compat

Comment (hwchase17) [reply langchain/vectorstores/pinecone.py:42]: lgtm... its backward compatible right?

Comment (dev2049) [reply langchain/vectorstores/pinecone.py:42]: afaict should be? unless someone was specifying it as a kwarg. can add a test to make sure

**Comment by baskaryan** [comment]
looks like we added some batching in #1658. closing for now, let me know if i'm missing something!

**Pull request closed without merging**